src/daft-context/src/lib.rs

```diff
@@ -0,0 +1,386 @@
+        use Runner::{Native, Py};
+        if let Some(current_runner) = self.runner() {
+            let runner = match (current_runner.as_ref(), runner.as_ref()) {
+                (Native(_), Native(_)) | (Py(_), Py(_)) => return Ok(()),
+                (Py(_), Native(_)) | (Native(_), Py(_)) => runner,
+
+                _ => {
+                    return Err(DaftError::InternalError(
+                        "Cannot set runner more than once".to_string(),
+                    ));
+                }
+            };
+
+            let mut state = self.state.write().map_err(|_| {
+                DaftError::InternalError("Failed to acquire write lock on DaftContext".to_string())
+            })?;
+
+            state.runner.replace(runner);
+
+            Ok(())
+        } else {
+            let mut state = self.state.write().map_err(|_| {
+                DaftError::InternalError("Failed to acquire write lock on DaftContext".to_string())
+            })?;
+
+            state.runner.replace(runner);
+            Ok(())
+        }
+    }
+
+    /// Get a read only reference to the state.
+    pub fn state(&self) -> RwLockReadGuard<'_, ContextState> {
+        self.state.read().unwrap()
+    }
+
+    /// get a mutable reference to the state.
+    pub fn state_mut(&self) -> std::sync::RwLockWriteGuard<'_, ContextState> {
+        self.state.write().unwrap()
+    }
+
+    /// get a read only reference to the catalog
+    pub fn catalog(&self) -> std::sync::MappedRwLockReadGuard<'_, DaftCatalog> {
+        RwLockReadGuard::map(self.state(), |state| &state.catalog)
+    }
+
+    /// get a mutable reference to the catalog
+    pub fn catalog_mut(&self) -> std::sync::MappedRwLockWriteGuard<'_, DaftCatalog> {
+        std::sync::RwLockWriteGuard::map(self.state_mut(), |state| &mut state.catalog)
+    }
+
+    /// get the execution config
+    pub fn execution_config(&self) -> Arc<DaftExecutionConfig> {
+        self.state().config.execution.clone()
+    }
+
+    /// get the planning config
+    pub fn planning_config(&self) -> Arc<DaftPlanningConfig> {
+        self.state().config.planning.clone()
+    }
+}
+
+#[cfg(not(feature = "python"))]
+impl DaftContext {
+    #[allow(clippy::new_without_default)]
+    pub fn new() -> Self {
+        unimplemented!()
+    }
+
+    pub fn get_or_create_runner(&self) -> DaftResult<Arc<Runner>> {
+        unimplemented!()
+    }
+
+    pub fn runner(&self) -> Option<Arc<Runner>> {
+        unimplemented!()
+    }
+
+    pub fn set_runner(&self, runner: Arc<Runner>) -> DaftResult<()> {
+        unimplemented!()
+    }
+
+    pub fn state(&self) -> RwLockReadGuard<'_, ContextState> {
+        unimplemented!()
+    }
+
+    pub fn state_mut(&self) -> std::sync::RwLockWriteGuard<'_, ContextState> {
+        unimplemented!()
+    }
+}
+
+static DAFT_CONTEXT: OnceCell<DaftContext> = OnceCell::new();
+
+#[cfg(feature = "python")]
+pub fn get_context() -> DaftContext {
```

Comment:
Currently we only have a singleton context as this is how it was done in py, but we may want to think about of ways that a user could initialize multiple contexts/sessions.

Reply:
hmm yeah I was actually thinking a bit about this yesterday as well. I think that seems like a reasonable distinction to make.

Reply:
Since we're going to make a hard distinction between the context and the session, I'll update the PR remove the `catalog` from the context.